async API

douglas-raillard-arm opened the pull request:
Add an async API to devlib to take advantage of concurrency more easily, both at a coarse and fine grain.

Note: This currently requires python 3.7

Based on https://github.com/ARM-software/devlib/pull/568

douglas-raillard-arm commented:
Preliminary results on:
```
        with target.cpufreq.use_governor('performance'):
            pass
```

30% speedup On a SSH target (juno r0):
    * on this branch: 3.7s (avg over 10 runs)
    * on master:        5.3s (avg over 10 runs)

50% speedup on a local target:
    * on this branch: ~9s 
    * on master:        ~20s

douglas-raillard-arm commented:
I checked all instances of threading.Lock in devlib and it should be safe, as none is in the path used when running a background command or any other coroutine.

If a function taking a threading.Lock was being called by 2 coroutines that end up running concurrently, the 2nd coroutine would wait for the lock, thereby blocking the main thread and deadlocking.

douglas-raillard-arm commented:
I also tried various combinations of thread pools, including running a blocking execute() in a different thread (so with separate SSH connections) and I did not get any real timing variations.

What I did get is some errors probably because I was maintaining too many connections on the SSH server, so the current approach with one connection and many concurrent SSH channels seems to be the best. I could not saturate the server with open channels (maybe there is no limit). If that was to happen, we could use an asyncio.Semaphore to limit the number of bg commands in flight.

douglas-raillard-arm commented:
There are a few things to cleanup (docstring etc) so if you are happy with the API as it stands I will proceed with:

    * adding docstrings
    * renaming asyn.parallel into asyn.concurrently
    * remove the .concurrent attribute, as it is not that useful in practice I think (it allows running the function concurrently, only blocking to get the result when it is awaited. The "normal" couroutine behavior is to not do anything until the return value is awaited). asyn.concurrently is usually better as it makes it obvious what runs concurrently and avoids any kind of "task leak".

douglas-raillard-arm commented:
@setrofim @marcbonnici I've updated the PR with an automatic detection of number of allowed background commands, along with some automatic limiting in the async API to use at most half of the available channels. 

Next step:
* Decide what version of Python we want to support
* "instrument" read_value() and write_value() such that we can detect when we are trying to read/write to the same files concurrently. This should catch most issues where we are trying to do some setup concurrently, which is where the bulk of the speed up will come from.

douglas-raillard-arm commented:
Updated the PR with checks to make sure that no file access conflicts when executing concurrent asyncio tasks. An artificial example:
```
from devlib.utils.asyn import *

m = ConcurrentResourceManager()

r1 = FileResource('host', 'file1', 'w')
r2 = FileResource('host', 'file1', 'r')

async def f1():
    # m.track_resource(r1)
    m.track_resource(r2)
    print('from f1')


async def f2():
    m.track_resource(r1)
    m.track_resource(r2)
    print('from f2')

async def f3():
    # m.track_resource(r1)
    print('from f3')

run(
    m.concurrently(
        (
            f1(),
            m.concurrently(
                (
                    f2(),
                    f3(),
                )
            )
        )
    )
)
```
In the PR, `ConcurrentResourceManager` is `target.resource_manager`, which is thread-local

douglas-raillard-arm commented:
PR updated with:
* forcing python 3.7 in setup.py
* Rename `ConcurrentResourceManager` to `AsyncManager`
* Almost all methods of `Target` have been converted. Not all of them have been "smartly" converted though, e.g. by doing things concurrently. 
* Some more modules/collector updated to make use of concurrent commands (cgroups module init)

douglas-raillard-arm commented:
@marcbonnici @setrofim 
Updated with a bunch of fixes that were needed to run on my test platform:
* Add `Target.execute(..., decode=True)` parameter. When `decode=False` is used, execute() returns  a tuple of bytes (stdout, stderr). This was necessary to successfully read sysfs content, which can contain binary data (from device tree). Note that the change to the Android connection class is entirely untested so a simple `Target.execute('echo hello; echo world >&2 ', decode=False)` on an android device would be useful.

* Cleanup the cgroup module. Trying to parallelize the setup shed some light on a few problems with the code structure. This PR migrates it to use cgroup v2, cleans some cruft and exposes more clearly what pieces will be retainable in a future cgroup API that allows full cgroup power and what will probably become deprecated.

douglas-raillard-arm commented:
I guess I could split the PR so we can merge the main bit, and then open a new one with the relatively unrelated fixes (cgroup v2 and read_tree_values() on binary content)

douglas-raillard-arm commented on devlib/utils/android.py:
```diff
@@ -549,19 +569,25 @@ def adb_shell(device, command, timeout=None, check_exit_code=False,
+    if output:
         try:
-            output, exit_code, _ = raw_output.replace('\r\n', '\n').replace('\r', '\n').rsplit('\n', 2)
+            output, exit_code, _ = output.replace(b'\r\n', b'\n').replace(b'\r', b'\n').rsplit(b'\n', 2)
```
I re-pushed the PR and that should be fixed now, thanks

marcbonnici replied:
Yep that's fixed, now however now it's getting stuck with: 
```
android.py in process_uuencode(x)
    549         # end
    550         if x:
--> 551             return codecs.decode(x, 'uu')
    552         else:
    553             return b''

TypeError: decoding with 'uu' codec failed (TypeError: a bytes-like object is required, not 'str')
```

douglas-raillard-arm replied:
Ah, I got the types swapped sorry about that, I pushed again

douglas-raillard-arm commented:
@marcbonnici PR updated with:

fixes to cgroups:
* only manipulate PIDs since thread support in cgroup v2 is significantly more complex and would require a dedicated API. This is probably best left for a future API
* Fix to `Controller.tasks('/')`. Since existing code apparently expects to list all the tasks in the root group (rather than the root of the subtree handled by the controller), handle it as a special case.

Fix to Target.install():
Add a `has_busybox=True` parameter to Connection.execute(), so that install() is able to install busybox without needing to run it (for Android). This is the WIP commit which will be squashed if it works well.

douglas-raillard-arm commented:
@marcbonnici @setrofim Updated the PR:
* turned Target.execute(decode=False) into a separate method `Target.execute_raw()`
* Implemented `Target.execute_raw()` on top of BackgroundCommand.communicate() to avoid invasive changes to Connection.execute().
* removed cgroups modification, as they would be better served by a new module
* Added `busybox --install -s` and the corresponding simplification in the code

marcbonnici reviewed (commented):
Apologies again for the delay in reviewing this.
 
I'm still looking through the code but have hit a few functional issues so far which I thought were worth mentioning at this point.

However mainly I'm having trouble getting the connection benchmark to complete on my devices (added comments), however I have not found a resolution yet. Have you seen this sort of behaviour on any of your devices?

douglas-raillard-arm commented on devlib/target.py:
```diff
@@ -1381,29 +1605,32 @@ def kick_off(self, command, as_root=False):
 
-    def ps(self, threads=False, **kwargs):
+    @asyn.asyncf
+    async def ps(self, threads=False, **kwargs):
```
Should I just split that in a separate PR for now then so we can focus first on the async bit ?

douglas-raillard-arm replied:
PR split at https://github.com/ARM-software/devlib/pull/585

douglas-raillard-arm commented on devlib/connection.py:
```diff
@@ -61,17 +62,128 @@ class ConnectionBase(InitCheckpoint):
+
+    @property
+    @memoized
+    def _max_bg(self):
```
We technically could but I think it's quite dangerous. If someone puts e.g. 15 when the real value is 10 "everything will work fine" until the client code runs a bunch of commands in parallel, at which point something bad will happen. It could be some sort of generic exception with 0 information to allow debugging, or possibly a misleading exception, or even why not just hanging. In any case, we cannot expect any indication of the actual problem and the location of the exception will be unhelpful.

What could be done is expose as a setting the max number of polled connection (possibly 0) so the user can reduce the time spent in the setup phase if they are ok with sequential code. This would be totally safe in all circumstances.

douglas-raillard-arm commented on devlib/connection.py:
```diff
@@ -61,17 +62,128 @@ class ConnectionBase(InitCheckpoint):
+
+        finally:
+            for bg in kill_list:
+                bg.__exit__(None, None, None)
```
Interesting, so this means it happened both on Juno (Linux distro I guess) and you nexus 10, so there must be some case not covered by how we kill the processes. I'll try to review the code and how it should be done. In the mean time, if you have a reproducer, would it be possible for you to open a shell on the board somehow and see the state of the process that is supposed to be killed using e.g. ps ?

marcbonnici replied:
So interestingly, again with your additional fix, on my emulator it's still getting stuck, but now on a seemly random process in the list (e.g. sometimes it is the 3rd, 5th etc.) when querying `ps` for said PID nothing is found, or for any other PID in the list however there do appear to be busybox tasks still running...

```
 grep BACKGROUND_COMMAND_UUID=577a502f173e41b5a594a58429123426); echo "\n$?"'
to_kill='-12479 -12487 -12495 -12503 -12512 -12520 -12528 -12536 -12544 -12552 -12560 -12568 -12576 -12584 -12592 -12600 -12608 -12616 -12624 -12632 -12640 -12648 -12656 -12664 -12672 -12680 -12688 -12696 -12704 -12712 -12720 -12728 -12736 -12744 -12752 -12760 -12768 -12776 -12784 -12792 -12800 -12808 -12816 -12824 -12832 -12840 -12848 -12856 -12864'
Waiting for bg.pid=12479
Waiting for bg.pid=12487
Waiting for bg.pid=12495
Waiting for bg.pid=12503
Waiting for bg.pid=12512
Waiting for bg.pid=12520
```

```
generic_x86:/ $ top | grep busybox
12881 shell        20   0 8.8M 2.2M 1.9M S  0.0   0.1   0:00.00 grep busybox
12543 shell        20   0 1.1M 4.0K    0 S  0.0   0.0   0:00.00 busybox sleep 31536000
12542 shell        20   0 9.1M 2.3M 1.9M S  0.0   0.1   0:00.00 sh -c sh -c '/data/local/tmp/bin/busybox sleep 31536000'
12527 shell        20   0 1.1M 4.0K    0 S  0.0   0.0   0:00.00 busybox sleep 31536000
12526 shell        20   0 9.1M 2.2M 1.9M S  0.0   0.1   0:00.00 sh -c sh -c '/data/local/tmp/bin/busybox sleep 31536000'
```

If I kill the remaining PIDs manually it unblocks but the original command returns an error code and looks like it couldn't find the problematic PIDs. Still looking into this... 

> Command 'adb emulator-5554 shell 'BACKGROUND_COMMAND_UUID=577a502f173e41b5a594a58429123426 sh -c '"'"'sh -c '"'"'"'"'"'"'"'"'/data/local/tmp/bin/busybox xargs kill -9 --'"'"'"'"'"'"'"'"'  '"'"'' 123 adb emulator-5554 shell 'BACKGROUND_COMMAND_UUID=577a502f173e41b5a594a58429123426 sh -c '"'"'sh -c '"'"'"'"'"'"'"'"'/data/local/tmp/bin/busybox xargs kill -9 --'"'"'"'"'"'"'"'"'  '"'"'' b'' b"kill: unknown pid '-12520'\nkill: unknown pid '-12536'\n"' returned non-zero exit status 123
```

douglas-raillard-arm replied:
Could there somewhat be a bug in `bg.pid` implementation ? If it were the case, `kill` would not try to kill the right processes, leaving some behind, which would then just block in `bg.__exit__` as all it does is close the streams and wait for the command to complete.

Note that getting the PID of a background command is not granted by the typical connection. None of adb or ssh supports that in the protocol, so we ended up having to grep ps output with this bg UUID. Maybe the update to `busybox ps` broke that somehow.

douglas-raillard-arm commented on devlib/connection.py:
```diff
@@ -61,17 +62,128 @@ class ConnectionBase(InitCheckpoint):
+                    '-{}'.format(bg.pid)
+                    for bg in kill_list
+                )
+                killer_bg.communicate(input=to_kill.encode())
```
When trying manually, was it with busybox, the system's kill command, or shell builtin ?

douglas-raillard-arm replied:
Ok so I guess the busybox implementation of kill can get confused between things like `-9` for the signal number and `-99` for PGIDs. Is there something else to fix on this command in the end ?

douglas-raillard-arm commented:
I've updated the PR with:
* fixes for missing () and await
* An (very unlikely) possible fix for the bg command benchmarking hang:
```
-            '{} xargs kill -9'.format(self.busybox),
+            '{} xargs kill -9 --'.format(self.busybox),
```

douglas-raillard-arm commented on devlib/connection.py:
```diff
@@ -61,17 +62,128 @@ class ConnectionBase(InitCheckpoint):
+                    '-{}'.format(bg.pid)
+                    for bg in kill_list
+                )
+                killer_bg.communicate(input=to_kill.encode())
```
Maybe it's due to an issue with `AdbBackgroundCommand.communicate()` then and its interaction with adb (missing newline at end of input or something like that)

marcbonnici replied:
That is possible, I'll take a closer look at this see if I can get it unstuck.

douglas-raillard-arm commented:
If that does not work, I'll try another strategy without any killer_bg, but instead blocking on some stdin for each command (rather than a plain sleep). This way I should be able to unblock them easily.

marcbonnici commented on devlib/connection.py:
```diff
@@ -61,17 +62,131 @@ class ConnectionBase(InitCheckpoint):
         self._close_lock = threading.Lock()
         self.busybox = None
+        self._bg_spawn_lock = threading.RLock()
+        self.logger = logging.getLogger(self.__name__)
```
Should this be `self.logger = logging.getLogger(self.__class__.__name__)`?

douglas-raillard-arm commented:
Updated PR. 
Added:
* Added `--` to `devlib.connection._kill_pgid_cmd` to avoid issues as well

Still to do:
* split the addition of busybox to PATH in a separate PR (and fix the issues with ps output)
* figure out why `bg.__exit__` is hanging on your setup

marcbonnici commented:
Finally found the issue on my android device and it was due to the PID reporting of adb connections. 
Occasionally it would report the PID of the command used to find the PID rather than the PID of the command itself due to PID reuse. 
I've submitted a [PR  ](https://github.com/ARM-software/devlib/pull/581)to fix this so can continue with on my android devices.

douglas-raillard-arm commented:
Very good news, so I'll try to shortly work on that PR to address the remaining things but I guess we don't have big blockers anymore so it shouldn't be a lot of work

douglas-raillard-arm commented:
PR updated:

Changes:
* Added max_bg connection parameter, so that users can force serialized execute() and also limit the auto-detection duration.
* split the PR for busybox symlinks so it's not blocking this one: https://github.com/ARM-software/devlib/pull/585

TODO:
* workaround/fix the android background command PID detection problem

douglas-raillard-arm commented:
PR updated again with:
* fix the android background command PID detection problem. 
This is somewhat redundant with https://github.com/ARM-software/devlib/pull/581 but tries another approach. Instead of printing the PID to stdout, it writes it to a temp file. This avoids the issues of streams redirection, that are acceptable for SSH since it needs to be done anyway, but for adb it would add a great deal of complexity.

douglas-raillard-arm commented on devlib/utils/android.py:
```diff
@@ -603,29 +604,28 @@ def adb_background_shell(conn, command,
-    command = "{} sh -c {}".format(uuid_var, quote(command))
+    pid_file = None
+    try:
+        pid_file = quote(conn.execute('mktemp').strip())
```
That's a bit problematic indeed. Maybe we could place the files in a subfolder of the devlib "home" on the target ? Then we just set TEMPDIR to point at that in that command.

douglas-raillard-arm commented:
@marcbonnici PR updated with a 2nd attempt to close the adb background PID race: the command is frozen and the thawed using SIGSTOP/SIGCONT. It's not pretty but is a relatively minor change on top of the existing code.

marcbonnici commented:
> @marcbonnici PR updated with a 2nd attempt to close the adb background PID race: the command is frozen and the thawed using SIGSTOP/SIGCONT. It's not pretty but is a relatively minor change on top of the existing code.

That's an interesting approach perhaps the way to go. Initially it seems to work well however occasionally something goes wrong when attempting to send the signals on my device.

```
while True:
    bg = ta.background("ls")
    print(bg.pid)
````
After a varying number of iterations it raises the following error:
```
    625 find_pid = f'''pid=$({conn.busybox} ps -A -o pid,args | {conn.busybox} {grep_cmd} | {conn.busybox} grep -v {quote(grep_cmd)} | {conn.busybox} awk '{{print $1}}') && kill -CONT "$pid" && printf "%s" "$pid"'''
    626 ps_out = conn.execute(find_pid)
--> 627 pid = int(ps_out)
    628 return (p, pid)

ValueError: invalid literal for int() with base 10: '/system/bin/sh: kill: : arguments must be jobs or process IDs\n'
```
So look like for some reason it's failing to find the correct PID, have you seen anything like this in your testing?

douglas-raillard-arm commented:
PR updated with speedup of the async path when non-blocking calls are actually not required (when just one asyncio task is in use).

Wrt to the `ps` output, I think we need to print the command and output (without the filtering of grep) to understand what is going on. Maybe `kill` simply rejects PIDs that are no longer alive (and therefore not really a PID) ? I can't think of how that could happen though, since the shell process is frozen so it should not finish until released.

douglas-raillard-arm commented:
I've tried a few thousand iterations of your test case and it worked without issues on my side. 

Also updated the PR to use `busybox kill` instead of the system's one, maybe that will help ?

douglas-raillard-arm commented:
Re-updated the PR with kill and printf swapped in PID detection. That will allow seeing the value of `$pid` even if `kill` fails for some reason.

marcbonnici commented:
Thanks for the update, running the same test shows that no PID is being detected
```
    624 # Find the PID and release the blocked background command with SIGCONT
    625 find_pid = f'''pid=$({busybox} ps -A -o pid,args | {busybox} {grep_cmd} | {busybox} grep -v {quote(grep_cmd)} | {busybox} awk '{{print $1}}') && {busybox} printf "%s" "$pid" && {busybox} kill -CONT "$pid"'''
--> 626 pid = int(conn.execute(find_pid))
    627 return (p, pid)

ValueError: invalid literal for int() with base 10: ''
```

That's interesting that you don't see the same problem, I'm seeing this problem on my Nexus 5 over an ADB connection, however I also tried it on a emulator which took a lot longer (left the above while loop running for a few minutes) but it did eventually hit the same error.

As you said I'll need to try printing the separate parts of the command to see where it is failing. I can only think that the command is exiting before the `kill -STOP` has a chance to execute for some reason but I would need to investigate further.

douglas-raillard-arm commented:
PR update with the `--` removed in the killer_bg command since it was causing issue for `busybox kill`. Also now use `busybox kill` rather than system's kill so that all platforms will behave the same.

douglas-raillard-arm commented:
Updated the PR with:
* some squashing of commits that were touching the same part of code
* A fix to `Target.get_connection()` that was not setting `conn.busybox` on new connections
* 2 new commits on top: Change the approach for non blocking execution

Previously, `Target._execute_async()` was based on `Target.background()`. This allowed sharing a single connection instance, but that also meant that it blocked until the command was spawned. Unfortunately, experiments have show that spawning a background command is quite expensive (e.g. on android where we end up with a blocking `ps`). Following Amdahl's law and experiments, that greatly reduced the speedup that was achievable (e.g. x2 speedup for 8 concurrent commands for one of my Juno SSH setup).

Fortunately, this is solvable: instead of relying on `Target.background()`, we can just create a thread pool with a separate connection instance each, and dispatch calls to these threads without blocking at all. This allows the best level of parallelism possible, and even avoids the inefficiencies of the `Target.background()` API (e.g. the need to know the PID even if it's not actually used). That also somewhat simplifies the implementation and separates it from any background command implementation bug.

Using this implementation, I was able to achieve a 15s runtime reduction on a LISA kernel test that takes 1min30 with devlib's master branch (it consists of varied uses of the API, lots of sysfs reading and writing, one background() execution for a test workload and some file pushing/pulling).

douglas-raillard-arm commented:
Updated PR:
* Split commits that are not required for this PR anymore into other PRs (fixes related to background())
* Squashed commits that are modifying `Target._execute_asyn()`

douglas-raillard-arm commented:
FWIW we have been using this PR in Lisa for a month now (Lisa repo vendorizes devlib as a git subtree) and so far I haven't observed or heard of any issue related to that

douglas-raillard-arm commented:
@marcbonnici PR updated with:
* Fix to `connect(max_async)`
* Fix to `cpufreq.use_governor()` racy file write
* Fix to `async.asyncf()` decorator for async generators: async generators are now consumed completely when crossing a blocking boundary, rather than failing to await on them. We could maybe do something a bit smarter to preserve lazyness provided that we can asyncio.run() anywhere, so I'll have a look later.
* Added  `asyn.memoized_method()` decorator that works for both async and non-async code. This does not memoize non-hashable data so it will avoid issues described here: https://github.com/ARM-software/devlib/issues/341

douglas-raillard-arm commented:
PR re-updated with a lazy async generator blocking shim. This preserves the laziness of the async gen, made possible by `nest_asyncio` package.
